pkg/policyfilter/rthooks/rthooks.go

```diff
@@ -92,15 +92,25 @@ func createContainerHook(_ context.Context, arg *rthooks.CreateContainerArg) err
 		return err
 	}
 
+	var containerFound bool
+	var container *corev1.ContainerStatus
 	namespace := pod.ObjectMeta.Namespace
+	pod, container, containerFound = arg.Watcher.FindContainer(containerID)
+	if !containerFound {
+		log.WithError(err).Warnf("failed to find container information %s, aborting hook.", containerID)
+	}
+
+	containerName := container.Name
```

Comment:
Thanks for pointing me to this code.

I had a look. There are ways to get the container name from the hook itself, which would be my preferred approach. That being said:
 i) this goes beyond the scope of this PR
 ii) the container name can be found in annotations which are not standard, which means that the hook may fail to figure out the container.
 
My suggestion would be to pass an empty containerName ("") for now. And then in the code that does the filtering based on the container name, we try  `arg.Watcher.FindContainer(containerID)` if we don't have the container name. The benefit of that is we only call FindContainer if the we have a container selector?

In a subsequent PR, we can extend the hook to also pass the container name.

What do you think?

Reply:
I think I got the idea; however, I would like to understand how to implement it correctly.

If I pass an empty container name to [pfState.AddPodContainer](https://github.com/cilium/tetragon/pull/2231/files#diff-0d06ff6b079b83e5948096c3ddab5693cdb637f590b5d6361a4ef84f2b244d6bR113), how can I use `arg.Watcher.FindContainer(containerID)` from [containerMatches](https://github.com/cilium/tetragon/pull/2231/files#diff-980e5be14cfb8e087d0d09a0f42b7219a9510ad91dd199221abdc98d1767bf34R221) (where the filtering logic is happening)? If I understand correctly, I also need to pass the `arg` struct from [createContainerHook](https://github.com/cilium/tetragon/blob/main/pkg/policyfilter/rthooks/rthooks.go#L34) to [pfState.AddPodContainer](https://github.com/cilium/tetragon/pull/2231/files#diff-0d06ff6b079b83e5948096c3ddab5693cdb637f590b5d6361a4ef84f2b244d6bR113) and probably store it somewhere (in `containerInfo`?) to use it when filtering because it holds the hook information for this container?
Maybe, an alternative would be to check if the container name is empty during filtering and if it is, try to find the pod (`corev1.Pod`) and the container in it by container ID? However, this sounds pretty complex and inefficient.
Sorry if I overcomplicate it, I just don't have a clear picture in my head at the moment :disappointed: 

Also, I realized I have a bug in rthooks [here](https://github.com/cilium/tetragon/pull/2231/commits/fc0f13cb6f5ee0257512f972ed3c59b9a491b508#diff-0d06ff6b079b83e5948096c3ddab5693cdb637f590b5d6361a4ef84f2b244d6bR100) because I don't return anything if the container was not found, I will fix it when making changes in this file.

-----------
And one more thing, I'm trying to set up a local environment to see OCI hooks in action and I followed the demo with [cri-o](https://github.com/cilium/tetragon/blob/main/contrib/rthooks/tetragon-oci-hook/docs/demo.md). Unfortunately, I get `connecting to agent (context deadline exceeded) failed: unix:///var/run/cilium/tetragon/tetragon.sock` logs in `/opt/tetragon/tetragon-oci-hook.log`. Just wanted to mention it because it affects my ability to see how these hooks work, I'm still figuring out what's the issue.

Reply:
> If I pass an empty container name to [pfState.AddPodContainer](https://github.com/cilium/tetragon/pull/2231/files#diff-0d06ff6b079b83e5948096c3ddab5693cdb637f590b5d6361a4ef84f2b244d6bR113), how can I use arg.Watcher.FindContainer(containerID) from [containerMatches](https://github.com/cilium/tetragon/pull/2231/files#diff-980e5be14cfb8e087d0d09a0f42b7219a9510ad91dd199221abdc98d1767bf34R221) (where the filtering logic is happening)? If I understand correctly, I also need to pass the arg struct from [createContainerHook](https://github.com/cilium/tetragon/blob/main/pkg/policyfilter/rthooks/rthooks.go#L34) to [pfState.AddPodContainer](https://github.com/cilium/tetragon/pull/2231/files#diff-0d06ff6b079b83e5948096c3ddab5693cdb637f590b5d6361a4ef84f2b244d6bR113) and probably store it somewhere (in containerInfo?) to use it when filtering because it holds the hook information for this container?
> Maybe, an alternative would be to check if the container name is empty during filtering and if it is, try to find the pod (corev1.Pod) and the container in it by container ID? However, this sounds pretty complex and inefficient.
> Sorry if I overcomplicate it, I just don't have a clear picture in my head at the moment 😞

Yap, you are right. Didn't think it through. Let's keep it simple then and leave it as you have it now.

> Also, I realized I have a bug in rthooks [here](https://github.com/cilium/tetragon/pull/2231/commits/fc0f13cb6f5ee0257512f972ed3c59b9a491b508#diff-0d06ff6b079b83e5948096c3ddab5693cdb637f590b5d6361a4ef84f2b244d6bR100) because I don't return anything if the container was not found, I will fix it when making changes in this file.

Thanks! Yeah, let's fail for now if we cannot find the container name.

> And one more thing, I'm trying to set up a local environment to see OCI hooks in action and I followed the demo with [cri-o](https://github.com/cilium/tetragon/blob/main/contrib/rthooks/tetragon-oci-hook/docs/demo.md). Unfortunately, I get connecting to agent (context deadline exceeded) failed: unix:///var/run/cilium/tetragon/tetragon.sock logs in /opt/tetragon/tetragon-oci-hook.log. Just wanted to mention it because it affects my ability to see how these hooks work, I'm still figuring out what's the issue.

Yap, it would be great to test it with the OCI hook, so let's fix it!

Are you using the following helm option as indicated in the doc:
>         --set tetragon.grpc.address="unix:///var/run/cilium/tetragon/tetragon.sock" \

the OCI hooks only work with a gRPC server listening on the unix socket.


Many thanks for helping out with this! I think it got more complicated that we originally expected, but we will figure it out :)

Reply:
Here's a PR to update the hooks side to support the container name: https://github.com/cilium/tetragon/pull/2248. It still needs some work (e.g, docs) but most of the things should be there.

Reply:
It took me some time but in the end, I managed to get a local environment with OCI hook setup! :partying_face: 
Initially, it didn't work because of the bug I mentioned before (no return statement when container is not found).

I still get `context deadline exceeded` (I used the options as indicated in the doc) and after that, I see this log `failCheck determined that we should not fail this container, even if there was an error`. However, it looks like this log is not related to adding containers because when that's the case I saw this log `failCheck determined that we should fail`.

The great thing is that I also see logs like `hook request to agent succeeded` when a new container is added. :tada: 
It turned out, we cannot use `arg.Watcher.FindContainer()` because there is no container ID inside the pod at the moment of checking the hook.
Thanks to your comment about annotations and after inspecting the request, I also realized we can get the name from there.

--------
Your PR looks great (but also pretty big :smile: ). I wonder what I should do in my PR, wait for yours to be merged and then use the name you added?
For now, I implemented a very similar logic as in [your commit](https://github.com/cilium/tetragon/pull/2248/commits/f8c5e73db291f16852e46b163fc72fbe6386412e) (I wanted to do it yesterday, but you were quicker:) ). I did it so everything could work from my PR too and it actually works!
```
time="2024-03-23T10:15:05Z" level=trace msg="policyfilter: add pod container" cgroup-id=6403 container-id=crio-6dfc31c8cc5405abaab3851c095fb0b3f590d64c3726144fd24a4815b9a6c846 container-name=ubuntu hook=create-container namespace=default pod-id=58397ab1-2fad-416e-908c-bfefc91c88d7 sub-system=rthook.policyfilter
```

--------
Also, I have a silly question: how do you usually clean up disk space used by tetragon images after multiple builds? Every time I build an image with `make image`, it takes about 2GB of disk space and I'm already afraid of building new images locally :sweat_smile:

Reply:
Thanks for the update.

> I still get context deadline exceeded (I used the options as indicated in the doc) and after that, I see this log failCheck determined that we should not fail this container, even if there was an error. However, it looks like this log is not related to adding containers because when that's the case I saw this log failCheck determined that we should fail.

This happens when the hook cannot contact the agent. The `failCheck` is just a check on whether the hook should fail the creation of the container or not. By default, the hook will _not_ fail containers in the `kube-system` namespace so that we can start tetragon  if the hook is running but the tetragon agent is not. 

> Your PR looks great (but also pretty big 😄 ). I wonder what I should do in my PR, wait for yours to be merged and then use the name you added?

That would me my preference. Would that be OK with you? You can rebase your PR on top of my branch until the PR is merged.

> I did it so everything could work from my PR too and it actually works!

Nice! :)

> Also, I have a silly question: how do you usually clean up disk space used by tetragon images after multiple builds? Every time I build an image with make image, it takes about 2GB of disk space and I'm already afraid of building new images locally 

Have you tried using `docker rmi` to remove the images?

Reply:
> That would me my preference. Would that be OK with you? You can rebase your PR on top of my branch until the PR is merged.
I'm absolutely fine with this! I can wait for it to be merged and then rebase and push the changes to avoid multiple rebasing:) (or you would like me to test if it works with my changes before merging?)

> Have you tried using docker rmi to remove the images?
I use `docker rmi $(docker images -f "dangling=true" -q)` to remove dangling images upon each build but it doesn't remove much:(
Even if I remove `cilium/tetragon:latest` locally, it doesn't clean up a lot of space :disappointed:

Reply:
> I just merged https://github.com/cilium/tetragon/pull/2248 so you can just rebase to main.

That's awesome, thank you!
I rebased to `main`, and now use `arg.Req.ContainerName` to get the container name. I tested it, it worked, and I pushed the changes:)

> We do use some images for the build, but I would expect those images to be matched by "dangling=true" as well. So not sure if there is a better way.

It was `buildx` cache! It was my first time using buildx in this project and it was caching layers of every image I built. I pruned it and now I'm good :relieved:

Reply:
Cool, thanks!

Please, "Re-request review" when you it's ready for review again.